entry/auth.cc

```diff
@@ -0,0 +1,140 @@
+#include "auth.h"
+
+std::string auth::gen_random(const int len) {
+    static const char alphanum[] =
+        "0123456789"
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        "abcdefghijklmnopqrstuvwxyz";
+    std::string tmp_s;
+    tmp_s.reserve(len);
+
+    for (int i = 0; i < len; ++i) {
+        tmp_s += alphanum[rand() % (sizeof(alphanum) - 1)];
+    }
+    
+    return tmp_s;
+}
+
+
+
+std::string auth::get_hash(std::string key){
+	unsigned char hash[SHA256_DIGEST_LENGTH];
+	SHA256_CTX sha256;
+	
+	SHA256_Init(&sha256);
+	SHA256_Update(&sha256, key.c_str(), key.size());
+	SHA256_Final(hash, &sha256);
+	
+	std::stringstream ss;
+	for(int i = 0; i < SHA256_DIGEST_LENGTH; i++){
+		ss << std::hex << std::setw(2) << std::setfill('0') << (int)hash[i];
+	}
+	return ss.str();
+}
+
+
+
+std::string auth::get_key(const crow::request& req){
+	std::string auth_header = req.get_header_value("Authorization");
+	std::string api_key_whole = auth_header.substr(7);
+	//std::string client_id = api_key_whole.substr(0, CLIENTIDLEN);
+	std::string api_key = api_key_whole.substr(CLIENTIDLEN, APIKEYLEN);
+	std::cout << api_key << std::endl;
+	return api_key;
+}
+
+
+
+std::string auth::get_client_id(const crow::request& req){
+	std::string auth_header = req.get_header_value("Authorization");
+	std::string api_key_whole = auth_header.substr(7);
+	std::string client_id = api_key_whole.substr(0, CLIENTIDLEN);
+	std::cout << client_id << std::endl;
+	return client_id;
+}
+
+
+
+bool auth::validate_credentials(const crow::request& req){
+	
+	std::string client_id = get_client_id(req);
+	std::string api_key = get_key(req);
+	
+	// get hash of apikey
+        std::string api_key_hash = get_hash(api_key);
+	
+	// TODO: Validate the credentials from db
```

Comment:
I think this can be a TODO for Jenny / whoever implements the db functionality cause they'll (hopefully) know more about the sqlite usage and can figure that out. If you wouldn't mind though, you can add to the interface in data.h and just call those. I also don't know if/how that'd affect the schema but shouldn't be difficult